src/esp/sensor/CameraSensor.cpp

```diff
@@ -115,20 +124,42 @@ bool CameraSensor::drawObservation(sim::Simulator& sim) {
   renderTarget().renderEnter();
 
   gfx::RenderCamera::Flags flags;
-  if (sim.isFrustumCullingEnabled())
+  if (sim.isFrustumCullingEnabled()) {
     flags |= gfx::RenderCamera::Flag::FrustumCulling;
+  }
 
-  gfx::Renderer::ptr renderer = sim.getRenderer();
   if (cameraSensorSpec_->sensorType == SensorType::Semantic) {
     // TODO: check sim has semantic scene graph
-    renderer->draw(*this, sim.getActiveSemanticSceneGraph(), flags);
-    if (&sim.getActiveSemanticSceneGraph() != &sim.getActiveSceneGraph()) {
+    bool twoSceneGraphs =
+        (&sim.getActiveSemanticSceneGraph() != &sim.getActiveSceneGraph());
+
+    if (twoSceneGraphs) {
+      // *assume* the sensor is always in the regular scene graph.
+      // backup first
+      Mn::Matrix4 relativeTransform = node().transformation();
+      Mn::Matrix4 absTransform = node().absoluteTransformation();
+      scene::SceneNode* p = static_cast<scene::SceneNode*>(node().parent());
+      // take the sensor from the current regular scene graph and connect it to
+      // the root node of semantic scene graph, set the *correct* transformation
+      node().setParent(&sim.getActiveSemanticSceneGraph().getRootNode());
+      node().setTransformation(absTransform);
+      draw(sim.getActiveSemanticSceneGraph(), flags);
+      // after drawing, put the node back to the regular scene graph.
+      // MUST use the setParent function! Otherwise the sensorSuite in the node
+      // will be wrong!!!
+      node().setParent(p);
+      node().setTransformation(relativeTransform);
+    } else {
+      draw(sim.getActiveSemanticSceneGraph(), flags);
+    }
+
+    if (twoSceneGraphs) {
       flags |= gfx::RenderCamera::Flag::ObjectsOnly;
```

Comment:
BTW, this cludge we are doing is going to hurt our FPS a lot in scenes without the semantic mesh, but only added objects (such as synthetic datasets.) One thing on the agenda is to find a better way to resolve this long term so we don't have to do this double pass all the time. @aclegg3 I think you mentioned something about this during our last meeting?

Reply:
This can of course be saved for a different PR, just wanted to bring it up.

Reply:
🤔  I'm not sure that this is the case. Here if we only have a single scene graph we'll just render it with semantic ids, if we have 2 scene-graphs (e.g. a specific separate semantic asset) then we render that asset and then a pass over dynamic objects. I don't see the overhead you are worried about. Am I missing something?

Reply:
@aclegg3 : It means the absolute transformation for all nodes will be computed twice (one per each SG).